### PR TITLE
Reproduce lecture 3 materials

### PR DESCRIPTION
**Pull request by khaletska**
In this PR I've recreated code from Lecture 3 of CS193P course (2013).
I also added my own style improvements to make it easier to read and to debug later if needed

### CODE REVIEW

File: Matchismo/Model/CardMatchingGame.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface CardMatchingGame : NSObject
+
+- (instancetype)initWithCardCounts:(NSUInteger)count usingDeck:(Deck*)deck NS_DESIGNATED_INITIALIZER;
+- (instancetype)init NS_UNAVAILABLE;
+- (void)chooseCardAtIndex:(NSUInteger)index;
+- (Card *)cardAtIndex:(NSUInteger)index;
+@property (nonatomic, readonly) NSInteger score;
```
Comment (Shiaulis) [inline Matchismo/Model/CardMatchingGame.h]: let's have all properties at the top of the interface and method at the bottom

File: Matchismo/Model/CardMatchingGame.m [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+//  Matchismo
+//
+//  Created by Yeliena Khaletska on 29.05.2024.
+//
+
+#import "CardMatchingGame.h"
+
+@interface CardMatchingGame()
+@property (nonatomic, readwrite) NSInteger score;
+@property (nonatomic, strong) NSMutableArray *cards;
```
Comment (Shiaulis) [inline Matchismo/Model/CardMatchingGame.m]: please specify the type

File: Matchismo/Model/CardMatchingGame.m [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+@end
+
+@implementation CardMatchingGame
+
+- (NSMutableArray *)cards
```
Comment (Shiaulis) [inline Matchismo/Model/CardMatchingGame.m]: please specify the type

File: Matchismo/Model/CardMatchingGame.m [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                }
+            }
+            self.score -= COST_TO_CHOOSE;
+            card.chosen = YES;
+        }
+    }
+}
+
+- (Card *)cardAtIndex:(NSUInteger)index {
+    return (index < self.cards.count) ? self.cards[index] : nil;
```
Comment (Shiaulis) [inline Matchismo/Model/CardMatchingGame.m:79]: can we use if-else instead?

File: Matchismo/Model/CardMatchingGame.m [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            }
+        }
+    }
+
+    return self;
+}
+
+static const int MISMATCH_PENALTY = 2;
+static const int MATCH_BONUS = 4;
+static const int COST_TO_CHOOSE = 1;
```
Comment (Shiaulis) [inline Matchismo/Model/CardMatchingGame.m]: let's declare all static properties on top of the file 🙏🏾

File: Matchismo/ViewController.m [hunk 6]
```diff
@@ -7,47 +7,60 @@
 
 #import "ViewController.h"
 #import "PlayingCardDeck.h"
+#import "CardMatchingGame.h"
 
 @interface ViewController ()
-@property (weak, nonatomic) IBOutlet UILabel *flipsLabel;
-@property (nonatomic) NSUInteger flipCount;
-@property (nonatomic) PlayingCardDeck *playingCardsDeck;
+@property (strong, nonatomic) IBOutletCollection(UIButton) NSArray *cardButtons;
```
Comment (Shiaulis) [inline Matchismo/ViewController.m]: please specify the type

File: Matchismo/ViewController.m [hunk 7]
```diff
@@ -7,47 +7,60 @@
...
 @implementation ViewController
 
-- (PlayingCardDeck *)playingCardsDeck
+- (CardMatchingGame *)game
 {
-    if (_playingCardsDeck == nil) {
-        _playingCardsDeck =  [PlayingCardDeck new];
+    if (_game == nil)
+    {
+        _game = [[CardMatchingGame alloc] initWithCardCounts:[self.cardButtons count] usingDeck:[self createDeck]];
```
Comment (Shiaulis) [inline Matchismo/ViewController.m]: count is a property

**Review by Shiaulis — CHANGES_REQUESTED** [review]
left couple of minor comments 🙏🏾 didn't test

File: Matchismo/Model/CardMatchingGame.m (hunk 4 above)
Comment (khaletska) [reply Matchismo/Model/CardMatchingGame.m:79]: we can, but why do we want to?

**Review by Shiaulis — APPROVED** [review]
great job 🙏🏾

**Pull request merged**